File: JsonSchemaAndStructs/KiteConnect-Haskell-Structs/instruments_nse.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE OverloadedStrings #-}

module InstrumentsNse
    ( InstrumentsNse (..)
    , InstrumentsNseElement (..)
    , Exchange (..)
    , InstrumentType (..)
    , decodeTopLevel
    ) where

import Data.Aeson
import Data.Aeson.Types (emptyObject)
import Data.ByteString.Lazy (ByteString)
import Data.HashMap.Strict (HashMap)
import Data.Text (Text)
import Data.Vector (Vector)

type InstrumentsNse = Vector InstrumentsNseElement

data InstrumentsNseElement = InstrumentsNseElement
    { exchangeInstrumentsNseElement :: Maybe Exchange
    , exchangeTokenInstrumentsNseElement :: Maybe Int
    , expiryInstrumentsNseElement :: Maybe Text
    , instrumentTokenInstrumentsNseElement :: Maybe Int
    , instrumentTypeInstrumentsNseElement :: Maybe InstrumentType
    , lastPriceInstrumentsNseElement :: Maybe Int
    , lotSizeInstrumentsNseElement :: Maybe Int
    , nameInstrumentsNseElement :: Maybe Text
    , segmentInstrumentsNseElement :: Maybe Exchange
    , strikeInstrumentsNseElement :: Maybe Int
    , tickSizeInstrumentsNseElement :: Maybe Float
    , tradingsymbolInstrumentsNseElement :: Maybe Text
    } deriving (Show)

data Exchange
    = NseExchange
    deriving (Show)

data InstrumentType
    = EqInstrumentType
    deriving (Show)

decodeTopLevel :: ByteString -> Maybe InstrumentsNse
decodeTopLevel = decode

instance ToJSON InstrumentsNseElement where
    toJSON (InstrumentsNseElement exchangeInstrumentsNseElement exchangeTokenInstrumentsNseElement expiryInstrumentsNseElement instrumentTokenInstrumentsNseElement instrumentTypeInstrumentsNseElement lastPriceInstrumentsNseElement lotSizeInstrumentsNseElement nameInstrumentsNseElement segmentInstrumentsNseElement strikeInstrumentsNseElement tickSizeInstrumentsNseElement tradingsymbolInstrumentsNseElement) =
        object
        [ "exchange" .= exchangeInstrumentsNseElement
        , "exchange_token" .= exchangeTokenInstrumentsNseElement
        , "expiry" .= expiryInstrumentsNseElement
        , "instrument_token" .= instrumentTokenInstrumentsNseElement
        , "instrument_type" .= instrumentTypeInstrumentsNseElement
        , "last_price" .= lastPriceInstrumentsNseElement
        , "lot_size" .= lotSizeInstrumentsNseElement
        , "name" .= nameInstrumentsNseElement
        , "segment" .= segmentInstrumentsNseElement
        , "strike" .= strikeInstrumentsNseElement
        , "tick_size" .= tickSizeInstrumentsNseElement
        , "tradingsymbol" .= tradingsymbolInstrumentsNseElement
        ]

instance FromJSON InstrumentsNseElement where
    parseJSON (Object v) = InstrumentsNseElement
        <$> v .:? "exchange"
        <*> v .:? "exchange_token"
        <*> v .:? "expiry"
        <*> v .:? "instrument_token"
        <*> v .:? "instrument_type"
        <*> v .:? "last_price"
        <*> v .:? "lot_size"
        <*> v .:? "name"
        <*> v .:? "segment"
        <*> v .:? "strike"
        <*> v .:? "tick_size"
        <*> v .:? "tradingsymbol"

instance ToJSON Exchange where
    toJSON NseExchange = "NSE"

instance FromJSON Exchange where
    parseJSON = withText "Exchange" parseText
        where
            parseText "NSE" = return NseExchange

instance ToJSON InstrumentType where
    toJSON EqInstrumentType = "EQ"

instance FromJSON InstrumentType where
    parseJSON = withText "InstrumentType" parseText
        where
            parseText "EQ" = return EqInstrumentType

File: JsonSchemaAndStructs/KiteConnect-Dart-Structs/instruments_nse.dart
// To parse this JSON data, do
//
//     final instrumentsNse = instrumentsNseFromMap(jsonString);

import 'package:freezed_annotation/freezed_annotation.dart';
import 'dart:convert';

part 'instruments_nse.freezed.dart';
part 'instruments_nse.g.dart';

@freezed
abstract class InstrumentsNse with _$InstrumentsNse {
    const factory InstrumentsNse({
        Exchange exchange,
        int exchangeToken,
        String expiry,
        int instrumentToken,
        InstrumentType instrumentType,
        int lastPrice,
        int lotSize,
        String name,
        Exchange segment,
        int strike,
        double tickSize,
        String tradingsymbol,
    }) = _InstrumentsNse;

    factory InstrumentsNse.fromJson(Map<String, dynamic> json) => _$InstrumentsNseFromJson(json);
}

enum Exchange { NSE }

final exchangeValues = EnumValues({
    "NSE": Exchange.NSE
});

enum InstrumentType { EQ }

final instrumentTypeValues = EnumValues({
    "EQ": InstrumentType.EQ
});

class EnumValues<T> {
    Map<String, T> map;
    Map<T, String> reverseMap;

    EnumValues(this.map);

    Map<T, String> get reverse {
        if (reverseMap == null) {
            reverseMap = map.map((k, v) => new MapEntry(v, k));
        }
        return reverseMap;
    }
}
